odoo/models.py

```diff
@@ -4283,24 +4283,26 @@ def _check_company(self, fnames=None):
         if not (regular_fields or property_fields):
             return
 
+        # If company_field is `0`, then record[company_field] will simply reference the current record
+        company_field = (
+            0 if self._name == 'res.company' else
+            'company_id' if 'company_id' in self._fields else
+            'company_ids' if 'company_ids' in self._fields else
+            None
+        )
+        if regular_fields and company_field is None:
+            _logger.warning(_(
+                "Skipping a company check for model %(model_name)s. Its fields %(field_names)s are set as company-dependent, "
+                "but the model doesn't have a `company_id` or `company_ids` field!",
+                model_name=self.model_name, field_names=regular_fields
+            ))
+
         inconsistencies = []
         for record in self:
             # The first part of the check verifies that all records linked via relation fields are compatible
             # with the company of the origin document, i.e. `self.account_id.company_id == self.company_id`
-            if regular_fields:
-                if self._name == 'res.company':
-                    companies = record
-                elif 'company_id' in self:
-                    companies = record.company_id
-                elif 'company_ids' in self:
-                    companies = record.company_ids
-                else:
-                    _logger.warning(_(
-                        "Skipping a company check for model %(model_name)s. Its fields %(field_names)s are set as company-dependent, "
-                        "but the model doesn't have a `company_id` or `company_ids` field!",
-                        model_name=self.model_name, field_names=regular_fields
-                    ))
-                    continue
+            if regular_fields and company_field is not None:
+                companies = record[company_field]
                 for name in regular_fields:
                     corecord = record.sudo()[name]
                     if corecord:
```

Comment:
maybe refactor in the future
I think `_check_company_domain` and `_check_company` should be consistent, if `_check_company_domain` handles model with `company_ids`, `_check_company` should also do that.